### ISSUE TITLE
Document "custom" measures added to PBIX with live SSAS tabular connection

### ISSUE THREAD

**Issue by wi11iamr** [DaxStudio/DaxStudio#113]
I understand that a live connection to an SSAS Tabular cube does not make the data model accessible within PowerBI and thus DAX Studio is unable to connect to it.
If this still holds true, are there any alternative approaches to using DAX Studio's MDSCHEMA_MEASURES DMV to document the additional measures that were created within the PBIX file itself?
(The reason for this is I use PBi to do dev/testing before then finalising the measures which will then be published to SSAS. Using DAX Studio to succinctly document all these new measures would help in identifying which development measures are not necessary to be published to the SSAS).

**Comment by dgosbell** [comment]
Your understanding is correct. When we connect to a model in Power BI Desktop are attempting connecting to the tables stored inside the pbix file. If you have created a report with a live connection to SSAS Tabular then there are no tables inside the pbix file itself.

Unfortunately all the features we have at present work against the data model, not against the report. So any measures created within the report itself are not visible. 

If you trace the queries that Power BI generates you would see that anytime you use one of these measures defined in the report, that Power BI injects the measure into the query with a DEFINE clause at the start of any query that references one of these measures.

I don't believe there are any public APIs that we can use to list these measures that are only defined in the report, but I'll leave this issue open and do some investigation to see if there is some way we can get at this information.

**Comment by erapade** [comment]
I strongly agree of the need to be able to connect DAX Studio to a Power BI Desktop report that has a live connection. Especially nowadays when it's possible to give users the Build role so they can connect to datasets even if they don't have workspace access.

Any news on this? Are there Microsoft dependencies required for you to be able to implement this and are there in that case any Power BI Idea logged that we can vote for?

**Comment by dgosbell** [comment]
@erapade - your request is different to the original one for this issue. DAX Studio connects to tabular datasets. When you have a report in live connect mode it does not contain any data. So all you have to do is to connect DAX Studio to the same dataset that your Live Connect Power BI report connects to. 

This works today if your live connection is to ASAzure or SSAS on-prem. It will work in future against Power BI Premium workspaces using the XMLA endpoint (which is currently in beta). In order to work against shared datasets that are not on a premium capacity would require that Microsoft extend the readonly  XMLA endpoint functionality to the shared capacities. I am not aware of anything on ideas.PowerBI.com for this since this feature is still in development and it has not been announced yet as to whether MS is planning to extend this to the shared capacities or not.

**Comment by erapade** [comment]
Thanks Darren (@dgosbell). Yes I understod that the request was different, I just thought that it would solve my issue as well, but that is probably a misunderstanding from my side.
If DAX Studio only can connect only to datasets as you say, the DAX Studio will never be able to connect to reports that does live connection to other datasets.

As I maybe didn't fully understand your answer, I will explain my scenarion so there is no misunderstanding
In my scenario I have a centrally managed dataset in a Power BI Premium supported workspace. I create reports towards this dataset, and within these reports I usually create new measures, and some of these measures are base measures for other measures. When optimizing, debugging, or developing new measures, I would like to be able to connect DAX Studio towards the live-connecting report (running on Power BI Desktop on my computer or maybe also in the Power BI Service), and then e.g. right-clicking on a measure and then choose "Define Dependent Measures".
Is this something that might be possible in the feature (i.e. will require a new idea post here), or something that will never be supported due to that DAX studio only will be able to connect to real datasets?

**Comment by dgosbell** [comment]
> If DAX Studio only can connect only to datasets as you say, the DAX Studio will never be able to connect to reports that does live connection to other datasets.

This is true. But as I mentioned, this is not a big issue as in most cases you can just connect DAX Studio to the same dataset as the report and everything just works.

> In my scenario I have a centrally managed dataset in a Power BI Premium supported workspace. I create reports towards this dataset, and within these reports I usually create new measures, and some of these measures are base measures for other measures. When optimizing, debugging, or developing new measures, I would like to be able to connect DAX Studio towards the live-connecting report (running on Power BI Desktop on my computer or maybe also in the Power BI Service), and then e.g. right-clicking on a measure and then choose "Define Dependent Measures".

So you can actually achieve the same outcome today, just through a slightly different means.

If you are already using a premium workspace you can [enable the xmla endpoint](https://powerbi.microsoft.com/en-us/blog/power-bi-open-platform-connectivity-with-xmla-endpoints-public-preview/)

Then you can build a visual using the measures you want to work with and use [this technique](https://www.sqlbi.com/articles/capturing-power-bi-queries-using-dax-studio/) and connect to the XMLA endpoint and capture the queries generated by Power BI. 

Although report level measures appear to be fully integrated into the metadata in reality they are not and Power BI has to inject the DEFINE clause for the measure and any other dependent measures into any query which references them at run time. 

It might be possible in future to import/browse measures from a live connect Power BI file (which is why I've left this issue open). But there is no official API from Microsoft for this, so unless they provide one we would need to reverse engineer the pbix file format and we probably would need to keep these measures separate from the normal metadata as you need to include them in a DEFINE block before you can reference them in a query.

**Comment by erapade** [comment]
Thanks Darren for explaining this. I do use the capture feature today, but I kind of didn't think of using it this way. A big thanks for this.

Concerning this: "_It might be possible in future to import/browse measures from a live connect Power BI file (which is why I've left this issue open). But there is no official API from Microsoft for this,..."
Do you have different channels to reach out to Microsoft with this, or shall there be placed an idea post to make this happening. If it requires a Idea post, I guess you are the best person to create it and if so, please plase a link (why not within DAX Studio itself :-) ) so we can vote for it

**Comment by dgosbell** [comment]
There is already an item on idea.powerbi.com for an [API to access pbix files](
https://ideas.powerbi.com/forums/265200-power-bi-ideas/suggestions/7345565-power-bi-designer-api) that has a lot of support. I do have access to some internal channels, but for large complex features like this one they need to prove significant customer impact (so they often suggest posting an item on ideas.powerbi.com)

**Comment by lisawid** [comment]
Although this is a super old thread, where are the measures of a live connected report then stored? I can't connect to the report without a premium subscription but the dataset does also not show me anything about them.
